lib/srv/alpnproxy/conn_test.go

```diff
@@ -0,0 +1,334 @@
+
+				// Expect routines to finish.
+				timer := time.NewTimer(10 * time.Second)
+				defer timer.Stop()
+				for i := 0; i < 1; i++ {
+					select {
+					case err := <-errChan:
+						require.NoError(t, err)
+					case <-timer.C:
+						require.Fail(t, "routing didn't finished in time")
+					}
+				}
+			})
+		}
+	})
+
+	t.Run("ConcurrentReads", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		nWrites := 10
+		dataWritten := []byte("message")
+
+		r, w := makePingConn(t)
+		defer r.Close()
+		defer w.Close()
+
+		readChan := make(chan []byte)
+
+		// Write routine
+		go func() {
+			for i := 0; i < nWrites; i++ {
+				_, err := w.Write(dataWritten)
+				if err != nil {
+					return
+				}
+			}
+		}()
+
+		// Read routines.
+		for i := 0; i < nWrites/2; i++ {
+			go func() {
+				buf := make([]byte, len(dataWritten)/2)
+				for {
+					n, err := r.Read(buf)
+					if err != nil {
+						return
+					}
+
+					chanBytes := make([]byte, n)
+					copy(chanBytes, buf[:n])
+					readChan <- chanBytes
+				}
+			}()
+		}
+
+	readLoop:
+		for i := 0; i < nWrites; i++ {
+			var aggregator []byte
+			for {
+				select {
+				case <-ctx.Done():
+					require.Fail(t, "failed to read message, expected '%s' but received '%s'", dataWritten, aggregator)
+				case data := <-readChan:
+					aggregator = append(aggregator, data...)
+					if bytes.Equal(dataWritten, aggregator) {
+						continue readLoop
+					}
+				}
+			}
+		}
+	})
+
+	t.Run("ConcurrentWrites", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		w, r := makeBufferedPingConn(t)
+		defer w.Close()
+		defer r.Close()
+
+		nWrites := 10
+		dataWritten := []byte("message")
+		writeChan := make(chan error)
+
+		// Start write routines.
+		for i := 0; i < nWrites/2; i++ {
+			go func() {
+				for writes := 0; writes < 2; writes++ {
+					err := w.WritePing()
+					if err != nil {
+						writeChan <- err
+						return
+					}
+
+					_, err = w.Write(dataWritten)
+					if err != nil {
+						writeChan <- err
+						return
+					}
+				}
+
+				writeChan <- nil
+			}()
+		}
+
+		// Expect all writes to succeed.
+		for i := 0; i < nWrites/2; i++ {
+			select {
+			case <-ctx.Done():
+				require.Fail(t, "timout write")
+			case err := <-writeChan:
+				require.NoError(t, err)
+			}
+		}
+
+		// Read all messages.
+		buf := make([]byte, len(dataWritten))
+		for i := 0; i < nWrites; i++ {
+			n, err := r.Read(buf)
+			require.NoError(t, err)
+			require.Equal(t, dataWritten, buf[:n])
+		}
+	})
+
+	t.Run("LargeContent", func(t *testing.T) {
+		if os.Getenv("TELEPORT_TEST_PING_CONN_LARGE") == "" {
+			t.Skip()
+		}
+
+		t.Run("SingleRead", func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+
+			r, w := makePingConn(t)
+			defer r.Close()
+			defer w.Close()
+
+			largeSlice := make([]byte, math.MaxUint32)
```

Comment:
We should not allocate 4GB in tests in any scenario. I don't think we should go over our `MaxHTTPRequestSize` for any reason: https://github.com/gravitational/teleport/blob/808be5be3b664d43cbf50f7ed55c3f90263feef4/constants.go#L500